doc/SAI_Proposal_Generic_Extensions.md

```diff
@@ -0,0 +1,423 @@
+attr.value.chardata[size - 1] = '\0';
+attrs.push_back(attr);
+ 
+// encode the JSON info for VIP table
+nlohmann::json json;
+ 
+// push metadata and value for virtual_ip
+nlohmann::json sai_metadata_json;
+sai_metadata_json[“sai_metadata”][“sai_attr_value_type”] = SAI_ATTR_VALUE_TYPE_IP_ADDRESS;
+nlohmann::json virtual_ip;
+virtual_ip[“virtual_ip”][“value”] = “10.1.1.1”;
+virtual_ip[“virtual_ip”][“sai_metadata”] = sai_metadata_json.dump();
+json.push_back(virtual_ip);
+ 
+// push metadata and value for nexthop_list
+nlohmann::json sai_metadata_json;
+sai_metadata_json[“sai_metadata”][“sai_attr_value_type”] = SAI_ATTR_VALUE_TYPE_OBJECT_LIST;
+nlohmann::json nexthop_list;
+nexthop_list [“nexthop_list”][“value”] = { nexthop_1_oid, nexthop_2_iod };
+nexthop_list [“nexthop_list”][“sai_metadata”] = sai_metadata_json.dump();
+json.push_back(nexthop_list);
+ 
+attr.id = SAI_GENERIC_PROGRAMMABLE_ATTR_ENTRY;
+strncpy(attr.data.json, json.dump(), sizeof(json.dump());
+attrs.push_back(attr);
+ 
+sai_object_id_t sai_generic_oid;
+sai_status_t status =
+    sai_generic_programmable_api->create_generic_programmable(
+        &sai_generic_oid, gSwitchId, (uint32_t)attrs.size(), attrs.data());
+```
+
+## Delete ##
+Delete requires no special information since the object is OID based. Just passing the OID should be sufficient to clear the entry.
+
+## Object Query ##
+The list of all generic programmable objects can be queried using the existing saiobject.h infrastructure in SAI.
+
+## CRM usage ##
+```c
+sai_attribute_t attr;
+vector<sai_attribute_t> attrs;
+ 
+// set object name
+std::string object_name = “vip_table”;
+attr.id = SAI_GENERIC_PROGRAMMABLE_ATTR_OBJECT_NAME;
+auto size = sizeof(attr.value.chardata);
+strncpy(attr.value.chardata, object_name.c_str(), size - 1);
+attr.value.chardata[size - 1] = '\0';
+attrs.push_back(attr);
+ 
+uint64_t count = 0;
+sai_status_t status =
+    sai_object_type_get_availability(gSwitchId,
+                                     SAI_OBJECT_TYPE_GENERIC_PROGRAMMABLE,
+                                     (uint32_t)attrs.size(),
+                                     attrs.data(),
+                                     &count);
+```
+
+# Usage Examples for sai_json_t #
+## Refer to an OID - Port ##
+```json
+{
+    "port_ref": {
+        "sai_metadata": {
+            "sai_attr_value_type": "SAI_ATTR_VALUE_TYPE_OBJECT_ID",
+            "allowed_object_types": ["SAI_OBJECT_TYPE_PORT"]
+        },
+        "value": "0x15000000000ffc"
+    }
+}
+```
+
+## Refer to an OID - Generic programmable block ##
+The sai_json_t type can include references to other SAI extension blocks. It is allowed to pass the new object type as a reference in the JSON like any other SAI object.
+```json
+{
+    "object_id": {
+        "sai_metadata": {
+            "sai_attr_value_type": "SAI_ATTR_VALUE_TYPE_OBJECT_ID",
+            "allowed_object_types": ["SAI_OBJECT_TYPE_GENERIC_PROGRAMMABLE"]
+        },
+        "value": "0x150000000008ba"
+    }
```

Comment:
How do we handle warm-upgrade?

1)  Say if the metadata of an attribute is changed across releases?

2) I introduce a new mandator attribute in the future release.

Reply:
My thinking is this behavior would mirror the existing SAI objects. The 2 different SAI implementations before and after warmboot should be able handle the difference in attributes.